src/Network/GRPC/MQTT/Sequenced.hs

```diff
@@ -118,8 +143,46 @@ mkPacketizedPublish client msgLimit topic = do
           modifyTVar' seqVar (+ 1)
           return $ toLazyByteString $ Packet isLastPacket curSeq (toStrict chunk)
 
-        liftIO $ publishq client topic rawPacket False QoS1 []
+        publishq client topic rawPacket False QoS1 []
 
         unless isLastPacket $ packetizedPublish rest
 
   return packetizedPublish
+
+data PublishToStream a = PublishToStream
+  { -- | A function to publish one data element.
+    publishToStream :: a -> IO ()
+  , -- | This function should be called to indicate that streaming is
+    -- completed.
+    publishToStreamCompleted :: IO ()
+  }
+
+mkStreamPublish :: forall r io. (Message r, MonadIO io) => Int64 -> (forall t. Message t => t -> IO ()) -> io (PublishToStream r)
```

Comment:
I don't think we can use this for all streaming calls by default. If I'm reading this right, the stream chunks are only flushed when the size limit is reached or `streamingDone` is called. This works well for endpoint using streaming to dump large amounts of data, such as a file download. But imagine a heartbeat endpoint, where the server is just sending a small message every 10 seconds. It is important that the message is actually sent at the time `publish` is called.

A quick and easy solution would be to provide a separate a separate flushing function that would allow the caller to force the messages to send without ending the stream. I think this would work, but it feels a bit counter-intuitive to me. It's batching by default, and the user needs to "opt-out" by doing something like `publishToStream >> flush` whenever they want to immediately send a message.

What make sense to me would be "opt-in" for batching. I don't have a clear idea of how it would be implemented, but some way for a user to mark an endpoint for batching, and then for those marked endpoints, we can use this batched publish.

Reply:
I added an option to "opt-in" to batching. This can be controlled via TH and/or proto file options. Details are added to README.